src/nsrt_learning/side_predicate_learning.py

```diff
@@ -455,70 +377,183 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
+            if not ground_op.preconditions.issubset(segment.init_atoms):
+                continue
+            # Check if effects match.
+            if not ground_op.add_effects.issubset(necessary_add_effects):
+                continue
+            return ground_op
+        return None
+
+    def _update_pnad_with_params_and_add_effects(
+            self, pnad: PartialNSRTAndDatastore, parameters: List[Variable],
+            add_effects: Set[LiftedAtom]) -> None:
+        """Update the given PNAD to change the parameters and add effects to
+        the given ones; also, update the preconditions.
+
+        Note that in general, changing the parameters means that we need
+        to recompute all datastores, otherwise the precondition learning
+        will not work correctly (since it relies on the substitution
+        dictionaries in the datastores being correct).
+        """
+        # Update the parameters and add effects of the operator. Set
+        # the preconditions to be trivial. They will be recomputed next.
+        pnad.op = pnad.op.copy_with(parameters=parameters,
+                                    preconditions=set(),
+                                    add_effects=add_effects)
+        # Recompute datastore.
+        self._recompute_datastores_from_segments([pnad],
+                                                 semantics="add_effects")
+        # Determine the preconditions.
+        preconditions = induce_pnad_preconditions(pnad)
+        # Update the preconditions of the operator.
+        pnad.op = pnad.op.copy_with(preconditions=preconditions)
+
+    @staticmethod
+    def _finalize_pnad_delete_effects(pnad: PartialNSRTAndDatastore) -> None:
+        """Update the given PNAD to change the delete effects to ones obtained
+        by unioning all lifted images in the datastore.
+
+        IMPORTANT NOTE: We do not allow creating new variables when we
+        do this. Instead, we filter out such delete effects from the
+        union. Therefore, even though it may seem on the surface like
+        this procedure will cause all delete effects in the data to be
+        modeled accurately, this is not actually true.
+        """
+        delete_effects = set()
+        for segment, var_to_obj in pnad.datastore:
+            objects = set(var_to_obj.values())
+            obj_to_var = {o: v for v, o in var_to_obj.items()}
+            atoms = {
+                atom
+                for atom in segment.delete_effects
+                if all(o in objects for o in atom.objects)
+            }
+            lifted_atoms = {atom.lift(obj_to_var) for atom in atoms}
+            delete_effects |= lifted_atoms
+        pnad.op = pnad.op.copy_with(delete_effects=delete_effects)
+
+    @staticmethod
+    def _finalize_pnad_side_predicates(pnad: PartialNSRTAndDatastore) -> None:
+        """Update the given PNAD to change the side predicates to ones that
+        include every unmodeled add or delete effect seen in the data."""
```

Comment:
for discussion later, would you agree with the following? side predicates serve two purposes:
1. there are some add effects that we would prefer not to model for generalization and planning efficiency reasons. side predicates give us a mechanism to ignore these add effects (this was their original purpose).
2. there are some delete effects that we _can't_ model, given the parameters that we have, so side predicates give us a way to throw these out anyway without actually including "deictic references" to new objects. we could instead introduce new parameters, but we don't want to, again for generalization and planning efficiency reasons.

Reply:
Exactly right! I fully agree and was planning to bring this up during our meeting today.

Reply:
yes, but we'll want to revise these claims after discussion later